keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,313 @@
+
+### Pubishing
+
+The `x-kubernetes-immutable` vendor extension is published via `/openapi/v2` as is.
+
+### Suggested marker syntax
+
+In analogy to `+required`, `+optional` we propose to add a marker to kubebuilder's
+controller-gen named `+immutable`, which covers the whole subtree of the object 
+(compare field-selection section for extension).
+
+```
+// The name can not be changed after creation.
+// +immutable
+Name string
+
+// The list of containers can not change AT ALL after creation.
+// No single field in existing containers can be changed, added or deleted,
+// no new containers can be added, no existing container can be removed.
+// +immutable
+Containers []Containers
+```
+
+### Future outline sketch: native resources
+
+For native resources we add a pre-immutability-check normalization step for objects 
+decoded from JSON which have normalizations defined:
+
+1. versioned JSON blob comes in a request
+2. unmarshalled into versioned Golang struct
+3. defaulting
+4. conversion to internal
+5. if immutability is enabled for the resource:
+   1. marshalling into JSON in-memory
+   2. normalization creating a copy with shared data-structures
+   3. strict immutability check against the old object, coming from proto assuming it is normalized   
+
+### Future outline sketch: field selection
+
+By default the equality is a strict deep-equal equality on a field. This can be generalized
+to an immutability mode. Here are examples:
+
+- **Default**: **Recursive**: means that none of their fields can change, and new fields are not accepted.
+- **Non-recursive**: For an array or a map, one can not remove or add new items, but existing items can be modified. 
+  For maps this means that the the key stay the same, but values might change. For arrays this means that the length
+  stays the same.
+- **Recursive with allowed addition and deletion**:
+  - New/deleted keys in maps are tolerated.
+  - ~~Appended items or removal of the last item in arrays are tolerated.~~
+    For arrays an identifier field is required, e.g. `.name`.
+- ~~**Mutable**: A field set as mutable could cancel the recursive immutability.~~ probably we don't want this.
+
+In CRDs these modes could be set via `x-kubernetes-immutable-mode: non-recursive/recursive/...`. Inside
+of Golang types the `+immutable` marker could be extended to `// +immutable=non-recursive/recursive/...`.
+
+More complex modes like allowed addition and removal with a custom identifier can be expressed with a
+multi key-value syntax:
+
+```
+// No single field in existing containers can be changed, added or deleted.
+// New containers can be added, existing containers can be removed. Containers
+// are identified by its name.
+// +immutable=addition=allowed,removal=allowed,identifier=.name
+Containers []Container
+```
+
+### Future outline: protobuf
+
+Protobuf encoding identifies unset, empty and null values for slices and maps. In native types
```

Comment:
The phrasing in this paragraph seems incompatible (or I'm just parsing it wrong) with what we say above with "the addition of protobuf or other encodings which unify values like empty, null and undefined.". Should this be "Protobuf encoding unifies..." too?

Reply:
identifies and unifies are the same in my understanding and this context. They are the same thing actually, i.e. cannot be differentiated. I can rephrase with unifies if you think that's clearer.

Reply:
It threw me, not sure how others will read it. But might switch to unify to be safe.

Reply:
fixed